web/static/js/components/simulation/QuestionnaireSimulation.jsx

```diff
@@ -87,21 +104,27 @@ class QuestionnaireSimulation extends Component<Props, State> {
     this.props.questionnaireActions.fetchQuestionnaireIfNeeded(projectId, questionnaireId)
   }
 
-  componentWillMount() {
-    const { projectId, questionnaireId, mode } = this.props
-    if (projectId && questionnaireId) {
-      this.fetchQuestionnaireForTitle()
-      if (['sms', 'ivr', 'mobileweb'].includes(mode)) {
-        startSimulation(projectId, questionnaireId, mode).then(result => {
-          this.setState({simulation: result})
-        })
-      }
+  // Callback for the mobileweb simulator running in an iframe.
+  onMessage(event) {
+    const { mode } = this.props
+    const { simulation } = this.state
+
+    if (event.data.simulationChanged && simulation) {
+      const { projectId, questionnaireId } = this.props
+      return fetchSimulation(projectId, questionnaireId, simulation.respondentId, mode).then(result => {
+        this.onSimulationChanged(result)
+      })
     }
   }
 
   onSimulationChanged = result => {
```

Comment:
I think we can re-write `onSimulationChanged`:
```
    const { simulation, isSimulationTerminated } = this.state
    const { t } = this.props

    if (isSimulationTerminated) {
      return
    }

    const { simulationStatus } = result
    const isSimulationTerminated = simulationStatus == 'expired' || simulationStatus == 'ended'

    if (isSimulationTerminated) {
      window.Materialize.toast(t(`This simulation is ${simulationStatus}. Please refresh to start a new one`))
    }

    let newSimulationState = {
      ...simulation,
      simulationStatus: simulationStatus
    }

    if (simulationStatus != 'expired') {
      // When the simulation expires we avoid refreshing (and so, erasing) the current state
      // So we only update the simulation status and we send a message to the end user
      // This allow us to handle this particular situation properly

      newSimulationState = {
        ...newSimulationState,
        messagesHistory: result.messagesHistory, // mode=sms
        prompts: result.prompts,                 // mode=ivr
        submissions: result.submissions,
        disposition: result.disposition,
        currentStep: result.currentStep
      }
    }

    this.setState({
      simulation: newSimulationState,
      isSimulationTerminated: isSimulationTerminated
    })
```
It's easy to see the definition of `simulation terminated` (and when to show the toast). Also it's easy to see when to update the complete simulation state (as stated in the comment). And last (but not least) the state is set in one place at the end of the function.

(note: I didn't try it, please verify this comment)

Reply:
You're right. That's much better.

Reply:
I only kept the translatable messages as-is, so they can still be extracted automatically, otherwise they'd get removed from the `translation.js` file.

Reply:
Great!